### PR TITLE
Mobile Benchmarks

### PR DESCRIPTION
**Pull request by david-cao**
This contains the code for the app used to benchmark both protobuf and gRPC on Android. The nano app used at first before switching over to lite in also included, just in case anyone wants to check it out. It only has protobuf benchmarks.

Also contains a Spark server for the HTTP side of things, and an incredibly simple python server if you're into that. 

In terms of review, the only part that really needs to be checked is the protolite_app section, @zsurocking  already had a quick passover of it. Other parts can be reviewed if you wish, i.e. Spark code, nano app. Also, making sure the readme is clear and understandable is important!

### CODE REVIEW

File: benchmarks-mobile/README.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
+Mobile gRPC Benchmarks
+======================
+These tools are what we use to benchmark mobile clients. Here you can find how to replicate our results and explore our results in more detail.
```
Comment (david-cao) [inline benchmarks-mobile/README.md]: Mention blog post here once it's up

File: benchmarks-mobile/README.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+Moreover, `HttpUrlConnection` does not seem to have a way to reuse the same object for multiple requests. Oracle mentions caching [here](https://docs.oracle.com/javase/6/docs/technotes/guides/net/http-keepalive.html), but it is unclear whether or not this is implemented in Android's SDK.
+
+There was a tiny difference between using `HttpURLConnection` and the OkHttp library, but it is almost negligible. OkHttp seems to be a bit faster for smaller messages, but worse for larger. You can test this yourself by checking the 'Use OkHttp' box. 
+
+A USB cable was used for our results in order to eliminate any discrepancies from using a wireless network. 
+
+Contributing
+------------
+If you have any thoughts on what we did, or wish to contribute your own benchmark results, please let us know! We're still looking for battery benchmarks, as there is no easy way to measure this.
```
Comment (david-cao) [inline benchmarks-mobile/README.md]: Any thing else we can add here?

**Comment by carl-mastrangelo** [comment]
@david-cao This PR is enormous and needs to be broken into smaller, review able pieces.  Please split this up into O(100)s of lines PRs.

Githubs review system is not suitable for a PR of this size.

**Comment by david-cao** [comment]
@carl-mastrangelo Ok, besides breaking up modular things, i.e. the two different apps, servers, readme, should I break up the app itself?

**Comment by carl-mastrangelo** [comment]
For starters, you could split this into a java and a boilerplate PR.  The boiler plate could have dummy java classes, that will be updated in the subsequent PR.

**Comment by ejona86** [comment]
I wouldn't have expected benchmark_results to be committed.

**Comment by david-cao** [comment]
Do you mean that the images should be hosted elsewhere or that the folder is unnecessary?

**Comment by ejona86** [comment]
@david-cao I'm surprised to see the folder. It doesn't seem like code or something that would be used by the app. It seems like results from one run.

**Comment by david-cao** [comment]
Yeah, they're there for the readme, should I move them elsewhere?

**Comment by ejona86** [comment]
@david-cao Ah, I see. They can be appropriate there then.

**Comment by david-cao** [comment]
See template at #2104

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @david-cao isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
CLA GitHub bot_

**Pull request closed without merging**